### ISSUE TITLE
Needs an `uninstall`

### ISSUE THREAD

**Issue by trajano** [corretto/amazon-corretto-crypto-provider#121]
When starting this up using a ServletContextListener and terminating the application it says that there are ThreadLocals that are still remaining.

**Comment by SalusaSecondus** [comment]
Our primary recommendation is to install ACCP at the server level rather than within a servlet or webapp. However, we will look into ways to address this.

Do you have any information on how other libraries have addressed this issue?

**Comment by trajano** [comment]
The best way I can think of is to create a webapp which loads up Cornetto as part of the servlet context
Then tell it to restart and see what is left over.

**Comment by SalusaSecondus** [comment]
`Security.removeProvider()` does uninstall the provider from the standard list. However it looks like all use of static `ThreadLocal`s are problems for this use-case. I'll keep investigating, but I'm not certain the best method yet.

**Comment by trajano** [comment]
I asked SO for some assitance on this https://stackoverflow.com/questions/63623208/how-do-i-determine-if-there-are-any-lingering-threadlocals-or-threads-on-java-wi?noredirect=1#comment112508343_63623208 

https://stackoverflow.com/questions/29269277/how-to-identify-and-remove-threads-threadlocals-initiated-from-our-webapp-in-jav/29641045#29641045

**Comment by SalusaSecondus** [comment]
I have a few ideas of ways to fix this by eliminating most of my static fields which chain to thread locals (and am investigating them). However, instantiating cryptographic providers of any time (not just ACCP) can be computationally expensive. You really should try to install it once at the server level and leave it installed and alone.

**Comment by trajano** [comment]
Yup I already have done that. It was more for the uncerntainty before From: SalusaSecondusSent: August 28, 2020 5:25 PMTo: corretto/amazon-corretto-crypto-providerCc: Archimedes Trajano; AuthorSubject: Re: [corretto/amazon-corretto-crypto-provider] Needs an `uninstall` (#121) I have a few ideas of ways to fix this by eliminating most of my static fields which chain to thread locals (and am investigating them). However, instantiating cryptographic providers of any time (not just ACCP) can be computationally expensive. You really should try to install it once at the server level and leave it installed and alone.—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by robin-aws** [comment]
Note I'm labelling this as "enhancement" given our recommendation against use cases that require uninstallation. It still may be better to use ThreadLocals in a friendlier way, but sounds like more investigation is still needed.
